connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -303,6 +292,20 @@ public void tick() {
         }
     }
 
+    private void processTargetStateChanges() {
+        if (!connectorTargetStateChanges.isEmpty()) {
```

Comment:
It looks like based on the synchronization in `onConnectorTargetStateChange` that you need synchronization around this, or you need to swap it out and save the copy locally in `tick()` like we do with `connectorConfigUpdatesCopy`.